src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(stateCopy, extraData);
+        break;
+
+      case 'removeProperties':
+        keyToRemove(stateCopy, keysToRemove);
+        break;
+
+      case 'clear':
+        keysToClear(stateCopy);
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
+  }
+
+  return stateHistory;
 }
 
+const keyToRemove = (state, action) => {
+  for (const key of action) {
+    delete state[key];
+  }
+};
+
+const keysToClear = (state) => {
+  for (const key in state) {
+    delete state[key];
+  }
+};
```

Comment:
better to move this function before `transformStateWithClones` - place where we use it